### ISSUE TITLE
 Surface Laptop 5 is missing in documentation

### ISSUE THREAD

**Issue by mbrennwa** [linux-surface/linux-surface#965]
As the title says, the new Surface Laptop 5 is not covered in the documentation (and in the feature matrix). Would be useful to include the SL5, even if all/many features are not yet supported or support is unknown.

**Comment by NP-chaonay** [comment]
I think this should be include Pro9/Studio2 also

btw it is just released, so stayed tune.

**Comment by qzed** [comment]
Right, haven't gotten around to adding the new devices yet, was a bit busy lately but I'll do that now.

**Comment by qzed** [comment]
I've added the entry for that, the Pro 9, and the Surface Studios. Please do let me know if there are any devices missing.

**Comment by mbrennwa** [comment]
What information / work is needed for full support of the Laptop 5 and the other new devices?

**Comment by qzed** [comment]
I'm not even sure what works and doesn't work at the moment. So most important right now: any sort of feedback. You can use the feature matrix as starting point.

And then ideally also a diagnostic dump or minimally an ACPI dump. See https://github.com/linux-surface/linux-surface/wiki/Testers-and-Contributing#support-for-new-devices (just added that part).

**Comment by mbrennwa** [comment]
Thanks. I will run these checks once I get my hands on a SL5.

**Comment by mnilsson76** [comment]
Hi,
I have a Surface Laptop 5 13.5" so here's the diagnostics.tar.gz for that. I've installed Ubuntu Desktop 22.10 and running the 6.0.7-surface kernel I can't get the keyboard, mousepad or touchscreen working.
[diagnostics.tar.gz](https://github.com/linux-surface/linux-surface/files/9994765/diagnostics.tar.gz)

**Comment by NP-chaonay** [comment]
OK so anyone here with SL5/SL4 

Could you do ithc touchscreen test, if this work on SL5 then it should work on SL4 too (but no one have test this yet)

```
1) install and  this driver instead of ipts driver
Github repo: quo/ithc recommended do on part "Building (DKMS)" and seeing "Notes" too
2) test the single touch, if it dont work then try using kernel paramenter intremap=nosid then retry
3) if it work then proceed to reinstall userspace driver from (at this moment) github linux-surface/iptsd  ; this can be either getting from prebuilt (more recommended) or  manual
4) after all then reboot then test the multitouch
```

**Comment by NP-chaonay** [comment]
in case driver test is sucess either single/multitouch

or in case it does not work both, but sure that it is using the same touch device as SP7+/SP8/SLS

you could able to report device and problem and info here

https://github.com/quo/ithc-linux/issues/5

**Comment by qzed** [comment]
@mnilsson76 Thanks! Can you try running the commands I posted here: https://github.com/linux-surface/linux-surface/issues/984#issue-1440834809?

**Comment by mnilsson76** [comment]
Hi again,
Right, I uninstalled iptsd and installed the ithc module and ran modprobe ithc, got this in dmesg:
`[   57.482433] ithc: loading out-of-tree module taints kernel.`

The touchscreen doesn't seem to work though, and neither does keyboard or mousepad. Also the module doesn't seem to autoload so I have to modprobe it manually.

Another weirdness with this thing is that everytime I boot I get grub twice, so I first choose Ubuntu, it goes black, the Windows logo comes for a second and then Grub reappears. When I choose Ubuntu a second time it boots. This happens also if I choose Windows.
I haven't disabled SecureBoot yet though...don't know if that has something to do with the Grub issues.

**Comment by mnilsson76** [comment]
> @mnilsson76 Thanks! Can you try running the commands I posted here: [#984 (comment)](https://github.com/linux-surface/linux-surface/issues/984#issue-1440834809)?

Sure, here you are, but as I said I have SL5 and not SP9

```
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/ctrl.py request 0x02 0x01 0x02 0x01 0x01
00 00 00 00 00 54 ba 00 00 e8 b7 00 00 01 00 00 00 9c 1d 00 00 00 00 00 00 00 00 00 00 04 00 00 00 e8 03 00 00 ff ff ff ff ff ff ff ff e8 03 00 00 e8 03 00 00 0a 00 00 00 0a 00 00 00 4d 31 30 38 37 32 37 33 00 00 00 00 00 00 00 00 00 00 00 00 00 30 31 36 36 38 31 36 32 31 39 00 4c 49 4f 4e 00 53 4d 50 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/ctrl.py request 0x02 0x01 0x0d 0x01 0x01
01 00 00 00

magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 0 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 1 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 2 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 3 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 4 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 5 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 6 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 7 0
[9, 33, 17, 1, 0, 1, 34, 109, 0]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 8 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 9 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 1 10 0
[]

magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 0 0
[9, 33, 17, 1, 0, 1, 34, 97, 3]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 1 0
[9, 33, 17, 1, 0, 1, 34, 128, 0]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 2 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 3 0
[9, 33, 17, 1, 0, 1, 34, 129, 1]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 4 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 5 0
[9, 33, 17, 1, 0, 1, 34, 78, 0]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 6 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 7 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 8 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 9 0
[]
magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/hid.py hid-get-descriptor 2 10 0
[]

```

**Comment by qzed** [comment]
Thanks!

> Sure, here you are, but as I said I have SL5 and not SP9

I am aware of that. Bring-up for the SAM/EC stuff (includes keyboard and touchpad) is similar across the newer Surface devices.

One more thing to test before I can start implementing a prototype: Can you also run the commands I posted in https://github.com/linux-surface/linux-surface/issues/984#issuecomment-1312459414? Also make sure you're on the latest surface kernel (v6.0.8-2) when running that.

**Comment by mnilsson76** [comment]
> Thanks!
> 
> > Sure, here you are, but as I said I have SL5 and not SP9
> 
> I am aware of that. Bring-up for the SAM/EC stuff (includes keyboard and touchpad) is similar across the newer Surface devices.
> 
> One more thing to test before I can start implementing a prototype: Can you also run the commands I posted in [#984 (comment)](https://github.com/linux-surface/linux-surface/issues/984#issuecomment-1312459414)? Also make sure you're on the latest surface kernel (v6.0.8-2) when running that.

Ok, here goes:

`magnus@sauron:~/surface-aggregator-module$ uname -a
Linux sauron 6.0.8-surface #2 SMP PREEMPT_DYNAMIC Sun Nov 13 10:53:13 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/evreg.py reg2-disable 0x15 0x00 0x03
00`

I got no events in dmesg as a result. Continuing:

`magnus@sauron:~/surface-aggregator-module$ sudo ./scripts/ssam/evreg.py reg2-enable 0x15 0x00 0x03
00
`

Now dmesg is spammed with unhandled events:
`[  281.010656] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x15, tc: 0x15, tid: 0x02, cid: 0x00, iid: 0x03)
[  281.016721] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x15, tc: 0x15, tid: 0x02, cid: 0x00, iid: 0x03)
[  281.024253] surface_serial_hub serial0-0: event: unhandled event (rqid: 0x15, tc: 0x15, tid: 0x02, cid: 0x00, iid: 0x03)
*snip*`

I get no more unhandled events after

`sudo ./scripts/ssam/evreg.py reg2-disable 0x15 0x00 0x03`

**Comment by qzed** [comment]
Perfect, thanks! I'll have something to test for you in a couple of minutes.

**Comment by qzed** [comment]
Okay, can you clone and test the `devices/sl5` branch of https://github.com/linux-surface/surface-aggregator-module?

To do that:
1. Clone the repo: `git clone https://github.com/linux-surface/surface-aggregator-module` and then `cd surface-aggregator-module`
2. Check out the `devices/sl5` branch: `git checkout devices/sl5`
3. Build the module: `cd module` and then `make all`
4. Load the module: `make insmod`

And with that, the keyboard, touchpad, and battery/charger status should hopefully work.

**Comment by mnilsson76** [comment]
Argh, SecureBoot is messing me up. I get:
```
insmod: ERROR: could not insert module surface_aggregator.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_aggregator_cdev.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_aggregator_hub.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_aggregator_tabletsw.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_acpi_notify.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_battery.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_charger.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_dtx.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_hid_core.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_hid.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_kbd.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_platform_profile.ko: Key was rejected by service
insmod: ERROR: could not insert module clients/surface_aggregator_registry.ko: Key was rejected by service
```

**Comment by qzed** [comment]
Ah right, the modules aren't automatically signed when building them. If your DKMS auto-signes the modules, you might be able to use `make dkms-install` instead of `make insmod` to permanently install them (use `make dkms-uninstall` to uninstall them again).

**Comment by mnilsson76** [comment]
Success! The keyboard, mousepad and battery/charging indicator started working!
Anything more you need from me, just ask! :)

**Comment by qzed** [comment]
Nice, thanks for the testing! I'll draft up a new release and once that is confirmed working I'll submit the patches upstream.

**Comment by qzed** [comment]
@mnilsson76 One thing I forgot to ask: Does the ACPI platform profile work? If you have the `surface-control` package installed, you can check this via the `surface profile` subcommand. In particular, try reading the profile (`surface profile get`), then setting it to something different (e.g. `surface profile set performance`, you can list all with `surface profile list`) and read it back again.

Alternatively, you can use the sysfs interface directly (read/write to `/sys/firmware/acpi/platform_profile`, read from `/sys/firmware/acpi/platform_profile_choices` for a list of available profiles).

**Comment by mnilsson76** [comment]
> @mnilsson76 One thing I forgot to ask: Does the ACPI platform profile work? If you have the `surface-control` package installed, you can check this via the `surface profile` subcommand. In particular, try reading the profile (`surface profile get`), then setting it to something different (e.g. `surface profile set performance`, you can list all with `surface profile list`) and read it back again.
> 
> Alternatively, you can use the sysfs interface directly (read/write to `/sys/firmware/acpi/platform_profile`, read from `/sys/firmware/acpi/platform_profile_choices` for a list of available profiles).

Yeah, it seems to work. I installed surface-control and can read/set it from there and also directly from Gnome. I have these profiles:
```
low-power
balanced
balanced-performance
performance
```

**Comment by qzed** [comment]
The profiles are fixed. As long as they stick when changing, then everything works. Thanks again!

**Comment by qzed** [comment]
@mnilsson76 New kernels are available in the repos. Can you verify that keyboard/touchpad and battery status work (if you have the surface-aggregator module installed via DKMS please make sure that this isn't being used/loaded, e.g. by uninstalling it)?

**Comment by NP-chaonay** [comment]
> The touchscreen doesn't seem to work though, and neither does keyboard or mousepad. Also the module doesn't seem to autoload so I have to modprobe it manually.

unless using amd version or haven't done intremap kernel cmd;  maybe there is some different of ithc (or using entirely different driver)

**Comment by mnilsson76** [comment]
> @mnilsson76 New kernels are available in the repos. Can you verify that keyboard/touchpad and battery status work (if you have the surface-aggregator module installed via DKMS please make sure that this isn't being used/loaded, e.g. by uninstalling it)?

Ok, done. I uninstalled the module through 'make dkms-uninstall', rebooted and verified that I no longer had keyboard/mousemap/ACPI. Updated from the repos, rebooted and it works again.

Regarding ACPI I seem to have the problem where a shutdown doesn't really shutdown but instead turns off the display and the machine keeps running, so I have to hold the power button to kill it.

**Comment by qzed** [comment]
@mnilsson76 Can you run `lsmod` and then `modprobe -r <module>` on all modules starting with `surface` and then try to shut down? Some modules might have dependencies, I think it should tell you about those, but you should confirm with `lsmod` again to check that you have unloaded all. Just want to make sure that the surface specific stuff isn't causing that issue.

**Comment by NP-chaonay** [comment]
@mnilsson76 

ok let me update

I think that above dont work for touchscreen

so could you give the `lspci -nn | grep -i digital | grep -i pen` output

Thanks Dorian Stoll for idea btw

**Comment by StollD** [comment]
They already posted the lspci output (part of the diagnostics script). The device ID is the same as on SP9 (its bound to the CPU generation not to the device)

**Comment by NP-chaonay** [comment]
ok then retry my comment https://github.com/linux-surface/linux-surface/issues/965#issuecomment-1312452135

(noted that pls reclone the ithc repo btw)

**Comment by qzed** [comment]
I've sent the SAM changes upstream: https://patchwork.kernel.org/project/platform-driver-x86/patch/20221115231440.1338142-1-luzmaximilian@gmail.com/. Should be included in v6.1

**Comment by qzed** [comment]
SAM (typecover, battery/charger-status, performance profiles, ...) stuff should be supported now. Touch I believe too with ITHC and IPTSd from the GitHub repos directly. There are still a bunch of question marks in the feature matrix.

I _think_ most if not all of that should work, so I'm closing this for now. If there are any remaining problems, please feel free to open up a new issue for those.
